Userland/Libraries/LibDeviceTree/FlattenedDeviceTree.cpp

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2023, Andrew Kaster <akaster@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/ByteBuffer.h>
+#include <AK/Error.h>
+#include <AK/IterationDecision.h>
+#include <AK/MemoryStream.h>
+#include <AK/StringView.h>
+#include <LibDeviceTree/FlattenedDeviceTree.h>
+
+namespace DeviceTree {
+
+static ErrorOr<StringView> read_string_view(ReadonlyBytes bytes, StringView error_str)
+{
+    auto len = strnlen(reinterpret_cast<char const*>(bytes.data()), bytes.size());
+    if (len == bytes.size()) {
+        return Error::from_string_view_or_print_error_and_return_errno(error_str, EINVAL);
+    }
+    return StringView { bytes.slice(0, len) };
+}
+
+ErrorOr<void> walk_device_tree(FlattenedDeviceTreeHeader const& header, ReadonlyBytes raw_device_tree, DeviceTreeCallbacks callbacks)
+{
+    ReadonlyBytes struct_bytes { raw_device_tree.data() + header.off_dt_struct, header.size_dt_struct };
+    FixedMemoryStream stream(struct_bytes);
+    char const* begin_strings_block = reinterpret_cast<char const*>(raw_device_tree.data() + header.off_dt_strings);
+
+    FlattenedDeviceTreeTokenType prev_token = EndNode;
+    StringView current_node_name;
+
+    while (!stream.is_eof()) {
+        auto next_token = TRY(stream.read_value<BigEndian<u32>>());
+
+        switch (next_token) {
+        case BeginNode: {
+            current_node_name = TRY(read_string_view(struct_bytes.slice(stream.offset()), "Non-null terminated name for FDT_BEGIN_NODE token!"sv));
+            size_t const consume_len = current_node_name.length() + 1 < 4 ? 4 : round_up_to_power_of_two(current_node_name.length() + 1, 4);
+            TRY(stream.discard(consume_len));
+            if (callbacks.on_node_begin) {
+                if (IterationDecision::Break == TRY(callbacks.on_node_begin(current_node_name)))
+                    return {};
+            }
+            prev_token = BeginNode;
```

Comment:
Could the prev_token just be assigned to `prev_token = next_token;` after the switch instead?

Reply:
Could, but since next_token is a BigEndian<u32> the static_casts get a bit messy.